File: apps/api/internal/http/assets.go
package http

import (
	"database/sql"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

// Vessel mirrors a row of vessel_positions.
type Vessel struct {
	MMSI      string     `json:"mmsi"`
	Name      *string    `json:"name"`
	ShipType  *string    `json:"ship_type"`
	Latitude  float64    `json:"latitude"`
	Longitude float64    `json:"longitude"`
	SOG       *float64   `json:"sog"`
	COG       *float64   `json:"cog"`
	Heading   *float64   `json:"heading"`
	Timestamp time.Time  `json:"timestamp"`
	Source    string     `json:"source"`
}

// Aircraft mirrors a row of aircraft_positions.
type Aircraft struct {
	ICAO24         string    `json:"icao24"`
	Callsign       *string   `json:"callsign"`
	OriginCountry  string    `json:"origin_country"`
	Latitude       float64   `json:"latitude"`
	Longitude      float64   `json:"longitude"`
	Altitude       *float64  `json:"altitude"`
	Velocity       *float64  `json:"velocity"`
	Heading        *float64  `json:"heading"`
	OnGround       bool      `json:"on_ground"`
	Timestamp      time.Time `json:"timestamp"`
}

const vesselsQuery = `
SELECT mmsi, name, ship_type, latitude, longitude,
       sog, cog, heading, timestamp, source
FROM vessel_positions
ORDER BY timestamp DESC
LIMIT 500
`

const aircraftQuery = `
SELECT icao24, callsign, origin_country, latitude, longitude,
       altitude, velocity, heading, on_ground, timestamp
FROM aircraft_positions
ORDER BY timestamp DESC
LIMIT 500
`

// AssetsMarine returns latest vessel positions.
func AssetsMarine(db *sql.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		if db == nil {
			c.JSON(http.StatusServiceUnavailable, gin.H{
				"error":   "database_unavailable",
				"message": "the database is not configured",
			})
			return
		}

		rows, err := db.QueryContext(c.Request.Context(), vesselsQuery)
		if err != nil {
			c.JSON(http.StatusServiceUnavailable, gin.H{
				"error":   "database_query_failed",
				"message": err.Error(),
			})
			return
		}
		defer rows.Close()

		vessels := make([]Vessel, 0, 100)
		for rows.Next() {
			var v Vessel
			if err := rows.Scan(
				&v.MMSI, &v.Name, &v.ShipType,
				&v.Latitude, &v.Longitude,
				&v.SOG, &v.COG, &v.Heading,
				&v.Timestamp, &v.Source,
			); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{
					"error":   "row_scan_failed",
					"message": err.Error(),
				})
				return
			}
			vessels = append(vessels, v)
		}

		if err := rows.Err(); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error":   "rows_iteration_failed",
				"message": err.Error(),
			})
			return
		}

		c.JSON(http.StatusOK, gin.H{
			"data": vessels,
			"meta": gin.H{"count": len(vessels)},
		})
	}
}

// AssetsAviation returns latest aircraft positions.
func AssetsAviation(db *sql.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		if db == nil {
			c.JSON(http.StatusServiceUnavailable, gin.H{
				"error":   "database_unavailable",
				"message": "the database is not configured",
			})
			return
		}

		rows, err := db.QueryContext(c.Request.Context(), aircraftQuery)
		if err != nil {
			c.JSON(http.StatusServiceUnavailable, gin.H{
				"error":   "database_query_failed",
				"message": err.Error(),
			})
			return
		}
		defer rows.Close()

		aircraft := make([]Aircraft, 0, 100)
		for rows.Next() {
			var a Aircraft
			if err := rows.Scan(
				&a.ICAO24, &a.Callsign, &a.OriginCountry,
				&a.Latitude, &a.Longitude,
				&a.Altitude, &a.Velocity, &a.Heading,
				&a.OnGround, &a.Timestamp,
			); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{
					"error":   "row_scan_failed",
					"message": err.Error(),
				})
				return
			}
			aircraft = append(aircraft, a)
		}

		if err := rows.Err(); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error":   "rows_iteration_failed",
				"message": err.Error(),
			})
			return
		}

		c.JSON(http.StatusOK, gin.H{
			"data": aircraft,
			"meta": gin.H{"count": len(aircraft)},
		})
	}
}

File: apps/worker/db/assets.py
"""Database operations for marine vessel and aircraft asset positions."""

from __future__ import annotations

import logging
from datetime import datetime, timezone

import asyncpg

from connectors.aisstream import VesselPosition
from connectors.opensky import AircraftPosition

logger = logging.getLogger(__name__)


async def upsert_vessels(pool: asyncpg.Pool, vessels: list[VesselPosition]) -> int:
    """Upsert vessel positions into vessel_positions table."""

    if not vessels:
        return 0

    rows = [
        (
            v.mmsi,
            v.name,
            v.ship_type,
            v.latitude,
            v.longitude,
            v.sog,
            v.cog,
            v.heading,
            v.nav_status,
            v.timestamp,
            v.source,
        )
        for v in vessels
    ]

    try:
        await pool.executemany(
            """
            INSERT INTO vessel_positions
                (mmsi, name, ship_type, latitude, longitude,
                 sog, cog, heading, nav_status, timestamp, source)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
            ON CONFLICT (mmsi) DO UPDATE SET
                name = EXCLUDED.name,
                ship_type = EXCLUDED.ship_type,
                latitude = EXCLUDED.latitude,
                longitude = EXCLUDED.longitude,
                sog = EXCLUDED.sog,
                cog = EXCLUDED.cog,
                heading = EXCLUDED.heading,
                nav_status = EXCLUDED.nav_status,
                timestamp = EXCLUDED.timestamp,
                source = EXCLUDED.source
            """,
            rows,
        )
        count = len(rows)
        logger.info("Upserted %d vessel positions", count)
        return count
    except Exception as e:
        logger.error("Failed to upsert vessels: %s", e)
        raise


async def fetch_vessels(pool: asyncpg.Pool, limit: int = 500) -> list[dict]:
    """Fetch latest vessel positions."""

    rows = await pool.fetch(
        """
        SELECT mmsi, name, ship_type, latitude, longitude,
               sog, cog, heading, timestamp, source
        FROM vessel_positions
        ORDER BY timestamp DESC
        LIMIT $1
        """,
        limit,
    )
    return [dict(r) for r in rows]


async def upsert_aircraft(pool: asyncpg.Pool, aircraft: list[AircraftPosition]) -> int:
    """Upsert aircraft positions into aircraft_positions table."""

    if not aircraft:
        return 0

    rows = [
        (
            a.icao24,
            a.callsign,
            a.origin_country,
            a.latitude,
            a.longitude,
            a.altitude,
            a.velocity,
            a.heading,
            a.on_ground,
            a.vertical_rate,
            a.timestamp,
            "opensky",
        )
        for a in aircraft
    ]

    try:
        await pool.executemany(
            """
            INSERT INTO aircraft_positions
                (icao24, callsign, origin_country, latitude, longitude,
                 altitude, velocity, heading, on_ground, vertical_rate,
                 timestamp, source)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)
            ON CONFLICT (icao24) DO UPDATE SET
                callsign = EXCLUDED.callsign,
                origin_country = EXCLUDED.origin_country,
                latitude = EXCLUDED.latitude,
                longitude = EXCLUDED.longitude,
                altitude = EXCLUDED.altitude,
                velocity = EXCLUDED.velocity,
                heading = EXCLUDED.heading,
                on_ground = EXCLUDED.on_ground,
                vertical_rate = EXCLUDED.vertical_rate,
                timestamp = EXCLUDED.timestamp,
                source = EXCLUDED.source
            """,
            rows,
        )
        count = len(rows)
        logger.info("Upserted %d aircraft positions", count)
        return count
    except Exception as e:
        logger.error("Failed to upsert aircraft: %s", e)
        raise


async def fetch_aircraft(pool: asyncpg.Pool, limit: int = 500) -> list[dict]:
    """Fetch latest aircraft positions."""

    rows = await pool.fetch(
        """
        SELECT icao24, callsign, origin_country, latitude, longitude,
               altitude, velocity, heading, on_ground, timestamp
        FROM aircraft_positions
        ORDER BY timestamp DESC
        LIMIT $1
        """,
        limit,
    )
    return [dict(r) for r in rows]
